File: go/stack.go
package main

type Stack[T any] []T

func NewStack[T any]() Stack[T] {
	return Stack[T]{}
}

func (s *Stack[T]) Push(v T) {
	*s = append(*s, v)
}

func (s *Stack[T]) Pop() {
	l := len(*s)
	if l == 0 {
		panic("empty stack")
	}

	*s = (*s)[:l-1]
}

func (s *Stack[T]) Size() int {
	return len(*s)
}

func (s *Stack[T]) Peek() T {
	l := len(*s)
	if l == 0 {
		panic("empty stack")
	}

	return (*s)[l-1]
}

func (s *Stack[T]) Iterate(f func(T)) {
	for i := len(*s) - 1; i >= 0; i-- {
		f((*s)[i])
	}
}

File: rust/src/bin/stack.rs
#![allow(dead_code)]

fn main() {}

struct Stack<T> {
    element: Vec<T>,
}

impl<T> Stack<T> {
    pub fn new() -> Stack<T> {
        Stack { element: Vec::new() }
    }

    pub fn push(&mut self, elem: T) {
        self.element.push(elem);
    }

    pub fn pop(&mut self) -> Option<T> {
        self.element.pop()
    }

    pub fn peek(&self) -> Option<&T> {
        self.element.last()
    }

    pub fn size(&self) -> usize {
        self.element.len()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn init() {
        let stack: Stack<i32> = Stack::new();
        assert_eq!(stack.size(), 0);
    }

    #[test]
    fn push() {
        let mut stack = Stack::new();
        stack.push(1);
        assert_eq!(stack.size(), 1);
    }

    #[test]
    fn pop() {
        let mut stack = Stack::new();
        stack.push(1);
        assert_eq!(stack.pop(), Some(1));
        assert_eq!(stack.size(), 0);
    }

    #[test]
    fn peek() {
        let mut stack = Stack::new();
        stack.push(1);
        assert_eq!(stack.peek(), Some(&1));
    }

    #[test]
    fn size() {
        let mut stack = Stack::new();
        stack.push(1);
        assert_eq!(stack.size(), 1);
    }
}
